File: flask_api/app/services/subscription_service.py
from datetime import timedelta
from app.dao.subscription_dao import SubscriptionDAO
from app.dao.membership_tier_dao import MembershipTierDAO
from app.dao.membership_plans_dao import MembershipPlansDAO
from app.services.tier_benefit_service import TierBenefitService
from flask import current_app

from app.constants import TierType
from app.utils import generate_subscription_number, calculate_end_date_from_current


class SubscriptionService:

    @staticmethod
    def get_all(page=1, per_page=10, filters=None):
        return SubscriptionDAO.get_all(page, per_page, filters)

    @staticmethod
    def get_by_id(subscription_id):
        return SubscriptionDAO.get_by_id(subscription_id)

    @staticmethod
    def create(user_id, data):
        plan_id = data["plan_id"]
        data["user_id"] = user_id
        current_app.logger.info(
            f"TierType:  {TierType.BRONZE.name} {TierType.BRONZE.value}"
        )
        bronze_tiers = MembershipTierDAO.get_active_by_tier(TierType.BRONZE.value)
        current_app.logger.info(
            f"bronze_tiers:  {[ tier.to_dict() for tier in bronze_tiers]} "
        )
        if len(bronze_tiers) > 0:
            default_tier_id = bronze_tiers[0].id

            benefits = TierBenefitService.get_by_tier_id(default_tier_id)
            serialized_benefits = [b.id for b in benefits]

            plan_details = MembershipPlansDAO.get_by_id(plan_id).to_dict()
            current_app.logger.info(
                f"default_tier_id : {default_tier_id} benefits : {benefits} plan_details: {plan_details}"
            )
            start_date, end_date = calculate_end_date_from_current(
                plan_details["duration_days"]
            )
            current_app.logger.info(f"start_date: {start_date} end_date: {end_date}")
            data["tier_id"] = default_tier_id
            data["current_tier"] = TierType.BRONZE.value
            data["benefits"] = serialized_benefits
            data["subscription_number"] = generate_subscription_number().upper()
            data["start_date"] = start_date
            data["end_date"] = end_date
            data["total_saved"] = 0
            data["total_spent"] = 0

        current_app.logger.info(f"Data: {data}")
        return SubscriptionDAO.create(data)

    @staticmethod
    def update(subscription_id, data):

        if data.get("tier_id"):
            benefits = TierBenefitService.get_by_tier_id(data.get("tier_id"))
            serialized_benefits = [b.id for b in benefits]
            data["benefits"] = serialized_benefits

        if data.get("plan_id"):
            plan_id = data["plan_id"]
            plan_details = MembershipPlansDAO.get_by_id(plan_id).to_dict()
            start_date, end_date = calculate_end_date_from_current(
                plan_details["duration_days"]
            )
            current_app.logger.info(f"start_date: {start_date} end_date: {end_date}")
            data["start_date"] = start_date
            data["end_date"] = end_date

        return SubscriptionDAO.update(subscription_id, data)

    @staticmethod
    def save(subscription):
        return SubscriptionDAO.save(subscription)

    @staticmethod
    def get_active_subscription_by_user(user_id):
        return SubscriptionDAO.get_active_by_user(user_id)

    @staticmethod
    def extend_subscription(subscription_id, extend_days=None, additional_data=None):
        subscription = SubscriptionDAO.get_by_id(subscription_id)
        if not subscription:
            return None

        if subscription.is_expired and not subscription.is_active:
            return None

        if extend_days:
            subscription.end_date += timedelta(days=int(extend_days))

        allowed_fields = ["auto_renew", "tier_id", "plan_id", "benefits"]
        for field in allowed_fields:
            if field in additional_data:
                setattr(subscription, field, additional_data[field])

        return SubscriptionDAO.save(subscription)

    @staticmethod
    def soft_delete(subscription_id):
        return SubscriptionDAO.soft_delete(subscription_id)

File: springboot_api/src/main/java/com/example/membership_api/service/SubscriptionService.java
package com.example.membership_api.service;

import java.time.LocalDateTime;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import com.example.membership_api.constants.TierType;
import com.example.membership_api.dao.MembershipPlanDAO;
import com.example.membership_api.dao.MembershipTierDAO;
import com.example.membership_api.dao.SubscriptionDAO;
import com.example.membership_api.model.MembershipPlan;
import com.example.membership_api.model.MembershipTier;
import com.example.membership_api.model.Subscription;
import com.example.membership_api.model.TierBenefits;
import com.example.membership_api.utils.DateUtils;
import com.example.membership_api.utils.JsonUtils;
import com.example.membership_api.utils.SubscriptionUtils;

import jakarta.transaction.Transactional;

@Service
@Transactional
public class SubscriptionService {

    @Autowired
    private SubscriptionDAO subscriptionDAO;

    @Autowired
    private MembershipTierDAO membershipTierDAO;

    @Autowired
    private MembershipPlanDAO membershipPlanDAO;

    @Autowired
    private TierBenefitService tierBenefitService;

    private static final Logger logger = LoggerFactory.getLogger(TierBenefitService.class);

    public Map<String, Object> getAll(int page, int perPage, Map<String, Object> filters) {
        return subscriptionDAO.getAll(page, perPage, filters);
    }

    public Subscription getById(String subscriptionId) {
        return subscriptionDAO.getById(subscriptionId);
    }

    public Subscription create(String userId, Map<String, Object> data) {
        String planId = (String) data.get("plan_id");

        logger.info("TierType: {} {}", TierType.BRONZE.name(), TierType.BRONZE.getLevel());

        List<MembershipTier> bronzeTiers = membershipTierDAO.getActiveByTier(TierType.BRONZE.getLevel());
        logger.info("bronze_tiers: {}", bronzeTiers.stream().map(MembershipTier::toMap).toList());
        
        Subscription subscription = new Subscription();
        
        if (!bronzeTiers.isEmpty()) {
            MembershipTier defaultTier = bronzeTiers.get(0);
            String defaultTierId = defaultTier.getId();

            var benefits = tierBenefitService.getByTierId(defaultTierId);
            String serializedBenefits = benefits.stream().map(b -> b.getId()).collect(Collectors.toList()).toString();

            MembershipPlan plan = membershipPlanDAO.getById(planId);
            logger.info("default_tier_id: {} benefits: {} plan_details: {}", defaultTierId, benefits, plan.toMap());

            LocalDateTime[] dates = DateUtils.calculateEndDateFromCurrent(plan.getDurationDays());
            LocalDateTime startDate = dates[0];
            LocalDateTime endDate = dates[1];

            subscription.setUserId(userId);
            subscription.setPlanId(planId);
            subscription.setTierId(defaultTierId);
            subscription.setCurrentTier(TierType.BRONZE.getLevel());
            subscription.setBenefits(serializedBenefits);
            subscription.setSubscriptionNumber(SubscriptionUtils.generateSubscriptionNumber().toUpperCase());
            subscription.setTotalSaved(0.0);
            subscription.setTotalSpent(0.0);
            subscription.setStartDate(startDate);
            subscription.setEndDate(endDate);
        }

        logger.info("subscription: {}", subscription.toString());
        return subscriptionDAO.create(subscription);
    }

    public Subscription update(String subscriptionId, Map<String, Object> data) {
        System.out.printf("Subscription Update: %s%n", data.toString());
        Map<String, Object> mutableData = (data instanceof HashMap) ? data : new HashMap<>(data);

        if (mutableData.containsKey("tierId")) {
            String tierId = mutableData.get("tierId").toString();
            System.out.printf("tierId: %s",tierId);
            List<TierBenefits> tierBenefits = tierBenefitService.getByTierId(tierId);            
                        List<String> serializedBenefits = tierBenefits.stream()
                .map(TierBenefits::getId)
                .collect(Collectors.toList());

            String jsonBenefits = JsonUtils.toJson(serializedBenefits); 

            // List<String> serializedBenefits = tierBenefits.stream()
            //     .map(b -> b.getId())
            //     .collect(Collectors.toList());
            
            System.out.printf("serializedBenefits: %s",jsonBenefits);
            mutableData.put("benefits", jsonBenefits);
        }

        if (mutableData.containsKey("planId")) {
            String planId = mutableData.get("planId").toString();
            MembershipPlan plan = membershipPlanDAO.getById(planId);
            LocalDateTime[] dates = DateUtils.calculateEndDateFromCurrent(plan.getDurationDays());
            mutableData.put("start_date", dates[0]);
            mutableData.put("end_date", dates[1]);
        }
        System.out.printf("mutableData: %s%n",mutableData.toString());
        return subscriptionDAO.update(subscriptionId, mutableData);
    }


    public Subscription save(Subscription subscription) {
        return subscriptionDAO.save(subscription);
    }

    public Subscription getActiveSubscriptionByUser(String userId) {
        return subscriptionDAO.getActiveByUser(userId);
    }

    public Subscription extendSubscription(String subscriptionId, Integer extendDays, Map<String, Object> additionalData) {
        Subscription subscription = subscriptionDAO.getById(subscriptionId);
        if (subscription == null || subscription.isExpired() && !subscription.getIsActive()) {
            return null;
        }

        if (extendDays != null) {
            subscription.setEndDate(subscription.getEndDate().plusDays(extendDays));
        }

        List<String> allowedFields = Arrays.asList("auto_renew", "tier_id", "plan_id", "benefits");
        for (String field : allowedFields) {
            if (additionalData != null && additionalData.containsKey(field)) {
                SubscriptionUtils.setField(subscription, field, additionalData.get(field));
            }
        }

        return subscriptionDAO.save(subscription);
    }

    public boolean softDelete(String subscriptionId) {
        return subscriptionDAO.softDelete(subscriptionId);
    }
}
